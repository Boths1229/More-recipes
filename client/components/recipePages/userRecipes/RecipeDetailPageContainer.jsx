import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes, { any } from 'prop-types';
import miniToastr from 'mini-toastr';
import { Lines } from 'react-preloading-component';
import Header from '../../common/Header';
import Footer from '../../common/Footer';
import upvoteRecipeAction from '../../../actionController/upvoteRecipe';
import downvoteRecipeAction from '../../../actionController/downvoteRecipe';
import favoriteRecipeAction from '../../../actionController/favoriteRecipe';
import getRecipeDetailsAction from '../../../actionController/getRecipeDetails';
import addReviewsAction from '../../../actionController/addReviews';
import RecipeDetailsCard from '../../Include/cards/RecipeDetailsCard';

/**
 * @class recipeDetailPageContainer
 *
 * @description container displaying recipe details and reviews
 */
class RecipeDetailPageContainer extends Component {
  /**
   *
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.state = {
      review: '',
    };
    this.upvoteRecipe = this.upvoteRecipe.bind(this);
    this.downvoteRecipe = this.downvoteRecipe.bind(this);
    this.favoriteRecipe = this.favoriteRecipe.bind(this);
    this.handleReviewChange = this.handleReviewChange.bind(this);
    this.addReview = this.addReview.bind(this);
    this.mainJsx = this.mainJsx.bind(this);
    this.errorJsx = this.errorJsx.bind(this);
    this.fetchingJsx = this.fetchingJsx.bind(this);
    this.renderJsx = this.renderJsx.bind(this);
  }

  /**
   * @description call the action to get a recipes
   *
   * @param {object} props
   *
   * @return {void} call getRecipeDetails action
   */
  componentDidMount() {
    const { recipeId } = this.props.match.params;
    if (!Number.isInteger(parseInt(recipeId, 10))) {
      this.props.history.push('/recipes');
    }

    if (this.props.authenticated) {
      return this.props.getRecipeDetailsAction(recipeId);
    }
    this.props.history.push('/recipes');
    miniToastr.init();
    return miniToastr.error('Login to continue');
  }

  /**
   * @description upvote a recipe
   *
   * @param {number} id id of recipe to be upvoted
   *
   * @return {void} calls upvoteRecipeAction
   */
  upvoteRecipe(id) {
    if (this.props.authenticated) {
      return this.props.upvoteRecipeAction(id);
    }
    miniToastr.init();
    return miniToastr.error('Login to continue');
  }

  /**
   * @description downvote a recipe
   *
   * @param {number} id id of recipe to be updated
   *
   * @return {void} calls downvoteRecipeAction
   */
  downvoteRecipe(id) {
    if (this.props.authenticated) {
      return this.props.downvoteRecipeAction(id);
    }
    miniToastr.init();
    return miniToastr.error('Login to continue');
  }

  /**
   * @description favorite a recipe
   *
   * @param {number} id - id of recipe to be favorited
   *
   * @return {void} calls favoriteRecipeAction
   */
  favoriteRecipe(id) {
    this.props.favoriteRecipeAction(id);
  }
  /**
   * @description set the state of value inputs on form
   *
   * @param {object} event
   *
   * @returns {void} set the state of value inputs on form
   */
  handleReviewChange(event) {
    const { name, value } = event.target;
    this.setState({
      [name]: value
    });
  }

  /**
   * @description component to display recipe details
   *
   * @returns {JSX} return JSX
   */
  mainJsx() {
    return (
      <div className="container">
        <RecipeDetailsCard
          recipeData={this.props.recipeDetails}
          upvoteRecipe={this.upvoteRecipe}
          downvoteRecipe={this.downvoteRecipe}
          favoriteRecipe={this.favoriteRecipe}
          addReview={this.addReview}
          value={this.state}
          reviews={this.props.reviews}
          onChange={this.handleReviewChange}
        />
        <div className="pb-5" />
      </div>
    );
  }

  /**
   * @description set the state of value inputs on form
   *
   * @returns {JSX} return JSX
   */
  fetchingJsx() {
    return (
      <div>
        <Lines />;
      </div>
    );
  }

  /**
   * @description component to be rendered when there is an error fetching recipe detail
   *
   * @returns {JSX} return JSX
   */
  errorJsx() {
    return (
      <div className="text-center error-margin">
        <div>
          <i className="fa fa-exclamation-circle fa-5x" />
        </div>
        <h1>{this.props.errorMessage}</h1>
      </div>
    );
  }

  /**
   * @description add a review
   *
   * @param {number} id - id of recipe to review
   *
   * @return {void} calls addReviews action
   */
  addReview(id) {
    const { review } = this.state;
    this.props.addReviewsAction(id, review);
  }

  /**
   * @description render recipe detail page based on input
   *
   * @returns {JSX} render JSX
   */
  renderJsx() {
    const { recipeDetailStatus } = this.props;
    if (recipeDetailStatus === 'fetched') return this.mainJsx();
    if (recipeDetailStatus === 'fetching') return this.fetchingJsx();
    if (recipeDetailStatus === 'error') return this.errorJsx();
    return null;
  }

  /**
   * @description render - display component
   *
   * @return {JSX} return JSX
   */
  render() {
    return (
      <div>
        <Header />
        { this.renderJsx() }
        <Footer />
      </div>
    );
  }
}

RecipeDetailPageContainer.defaultProps = {
  recipeDetailStatus: null,
  reviews: undefined
};

RecipeDetailPageContainer.propTypes = {
  getRecipeDetailsAction: PropTypes.func.isRequired,
  upvoteRecipeAction: PropTypes.func.isRequired,
  downvoteRecipeAction: PropTypes.func.isRequired,
  favoriteRecipeAction: PropTypes.func.isRequired,
  recipeDetails: PropTypes.objectOf(any).isRequired,
  match: PropTypes.objectOf(any).isRequired,
  authenticated: PropTypes.bool.isRequired,
  addReviewsAction: PropTypes.func.isRequired,
  errorMessage: PropTypes.string.isRequired,
  recipeDetailStatus: PropTypes.string,
  history: PropTypes.objectOf(any).isRequired,
  reviews: PropTypes.arrayOf(any)
};

const mapStateToProps = state => ({
  recipeDetails: state.getRecipeDetailsReducer,
  errorMessage: state.getRecipeDetailsReducer.errorMessage,
  recipeDetailStatus: state.getRecipeDetailsReducer.recipeDetailStatus,
  authenticated: state.authReducer.isAuthenticated,
  reviews: state.getRecipeDetailsReducer.recipeDetails.Reviews
});

const mapDispatchToProps = dispatch => (
  bindActionCreators({
    getRecipeDetailsAction,
    upvoteRecipeAction,
    downvoteRecipeAction,
    favoriteRecipeAction,
    addReviewsAction
  }, dispatch)
);

export default connect(mapStateToProps, mapDispatchToProps)(RecipeDetailPageContainer);
