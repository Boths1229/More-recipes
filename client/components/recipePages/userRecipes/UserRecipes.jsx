import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes, { any } from 'prop-types';
import { bindActionCreators } from 'redux';
import { isEmpty } from 'lodash';
import { Lines } from 'react-preloading-component';
import ReactPaginate from 'react-paginate';
import RecipeNotFound from '../../Error/RecipeNotFound';
import getUserRecipe from '../../../actionController/getUserRecipe';
import addRecipeAction from '../../../actionController/addRecipe';
import saveImageToCloudAction from '../../../actionController/saveImageToCloud';
import AddRecipeButton from '../../Include/buttons/AddRecipeButton';
import deleteRecipeAction from '../../../actionController/deleteRecipe';
import editRecipeAction, { setEditRecipeIdAction } from '../../../actionController/editRecipe';
import Header from '../../common/Header';
import Footer from '../../common/Footer';
import RecipeGrid from '../RecipeGrid';

/**
 * @class UserRecipes
 *
 * @description UserRecipes
 */
class UserRecipes extends Component {
  /**
   * @description create an instance of UserRecipes
   *
   * @param {props} props
   */
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      description: '',
      ingredient: '',
      recipeData: [],
      isFetched: false,
      isChanged: false,
      responseMessage: '',
    };
    this.handleChange = this.handleChange.bind(this);
    this.addRecipe = this.addRecipe.bind(this);
    this.deleteRecipe = this.deleteRecipe.bind(this);
    this.saveImageToCloud = this.saveImageToCloud.bind(this);
    this.editRecipe = this.editRecipe.bind(this);
    this.handlePaginationChange = this.handlePaginationChange.bind(this);
    this.renderRecipeGrid = this.renderRecipeGrid.bind(this);
    this.setEditRecipeId = this.setEditRecipeId.bind(this);
  }
  /**
   * @description check the state of isFetched and call the get recipe action
   *
   * @param {props} props
   *
   * @returns {void} return all recipes
   */
  componentDidMount() {
    const { userId } = this.props;
    const page = 0;
    this.props.getUserRecipe(userId, page);
  }

  /**
   * @description update the state of user recipes
   *
   * @param {object} nextProps
   *
   * @return {void} updated user recipe
   */
  componentWillReceiveProps(nextProps) {
    const { recipeImageUrl, recipes, addRecipeResponse } = nextProps;
    const { recipeData, isFetched } = recipes;
    if (!isEmpty(addRecipeResponse)) {
      this.setState({
        responseMessage: nextProps.addRecipeResponse
      });
    }
    if (recipeData !== this.props.recipes.recipeData) {
      this.setState({
        recipeData,
        isFetched
      });
    }

    if (recipeImageUrl !== this.props.recipeImageUrl) {
      this.setState({
        image: recipeImageUrl
      });
    }
  }

  /**
   * @description get the id of recipe to be edited
   *
   * @param {number} recipeId
   *
   * @returns {void} call setEditRecipeId action
   *
   */
  setEditRecipeId(recipeId) {
    this.props.setEditRecipeIdAction(recipeId);
  }

  /**
   * @description function to edit a recipe
   *
   * @param {number} id id of recipe to be edited
   *
   * @returns {void} calls editRecipeAction
   */
  editRecipe(id) {
    const { name, description, ingredient } = this.state;
    this.props.editRecipeAction(id, { name, description, ingredient });
  }

  /**
   * @description function to delete a recipe
   *
   * @param {number} id id of recipe to be deleted
   *
   * @returns {void} calls the delete recipe action
   */
  deleteRecipe(id) {
    this.props.deleteRecipeAction(id);
  }

  /**
   * @description set the state of value inputs on form
   *
   * @param {object} event
   *
   * @returns {void} set the state of value inputs on form
   */
  handleChange(event) {
    const { name, value } = event.target;
    this.setState({
      [name]: value,
      isChanged: true
    });
  }

  /**
   * @description function to add a recipe
   *
   * @param {object} event
   *
   * @returns {void} calls add recipe action
   */
  addRecipe(event) {
    event.preventDefault();
    const {
      name, image, description, ingredient
    } = this.state;
    const newRecipe = {
      name, image, description, ingredient
    };
    this.props.addRecipeAction(newRecipe);
    this.setState({
      name: '', image: '', description: '', ingredient: ''
    });
  }

  /**
   * @description upload Image to cloud
   *
   * @param {object} event
   *
   * @returns {void} call saveImageToCloudAction to save images
   */
  saveImageToCloud(event) {
    const image = event.target.files[0];
    if (image) {
      this.props.saveImageToCloudAction(image);
    }
  }


  /**
   * @description get user recipes to be displayed on the new page
   *
   * @param {object} recipes
   *
   * @return {void} calls getUserRecipe action
   */
  handlePaginationChange(recipes) {
    const { userId } = this.props;
    this.props.getUserRecipe(userId, recipes.selected);
  }

  /**
   * @description get user recipes to be displayed on the new page
   *
   * @param {object} recipes
   *
   * @return {void} calls getUserRecipe action
   */
  renderRecipeGrid() {
    return (
      <div className="row">
        { this.props.recipes.length === 0 ? <RecipeNotFound /> :
        <RecipeGrid
          recipes={this.props.recipes}
          setEditRecipeId={this.setEditRecipeId}
          showActionButton
          recipeData={this.state.recipeData}
          deleteRecipe={this.deleteRecipe}
          editRecipe={this.editRecipe}
          addRecipe={this.addRecipe}
          value={this.state}
          handleCloseModal={this.handleCloseModal}
          onChange={this.handleChange}
          saveImageToCloud={this.saveImageToCloud}
          {...this.state}
        />
      }
      </div>
    );
  }
  /**
   * @description render user recipes
   *
   * @returns {JSX} JSX
   */
  render() {
    return (
      <div>
        <Header />
        <div className="container">
          <div className="top-nav-bar">
            <AddRecipeButton
              value={this.state}
              onChange={this.handleChange}
              addRecipe={this.addRecipe}
              saveImageToCloud={this.saveImageToCloud}
              recipeImage={this.props.recipeImageUrl}
            />

          </div>
          <div>
            {
              this.props.recipes.isFetching ? <Lines /> :
              this.renderRecipeGrid()

            }
          </div>
          <div className="pt-3 pb-5">
            <ReactPaginate
              previousLabel="Previous"
              nextLabel="Next"
              breakLabel={<a href="">...</a>}
              breakClassName="page-link"
              onPageChange={this.handlePaginationChange}
              pageCount={this.props.page}
              containerClassName="pagination justify-content-center"
              pageLinkClassName="page-link"
              nextLinkClassName="page-link"
              previousLinkClassName="page-link"
              disabledClassName="disabled"
              pageClassName="page-item"
              previousClassName="page-item"
              nextClassName="page-item"
              activeClassName="active"
              subContainerClassName="pages pagination"
            />
          </div>
          <Footer />
        </div>
      </div>
    );
  }
}
UserRecipes.propTypes = {
  getUserRecipe: PropTypes.func.isRequired,
  addRecipeAction: PropTypes.func.isRequired,
  deleteRecipeAction: PropTypes.func.isRequired,
  editRecipeAction: PropTypes.func.isRequired,
  recipes: PropTypes.arrayOf(any).isRequired,
  recipeImageUrl: PropTypes.string.isRequired,
  saveImageToCloudAction: PropTypes.func.isRequired,
  userId: PropTypes.number.isRequired,
  addRecipeResponse: PropTypes.string.isRequired,
  setEditRecipeIdAction: PropTypes.func.isRequired,
  page: PropTypes.number.isRequired
};

/**
 * @description allow state to be available to UserRecipes class as props
 *
 * @param {state} state
 *
 * @returns {object} object
 */
const mapStateToProps = state => ({
  recipes: state.getUserRecipeReducer.recipeData,
  userId: state.authReducer.userData.id,
  page: state.getUserRecipeReducer.page,
  recipeImageUrl: state.saveImageToCloud.image,
  addRecipeResponse: state.addRecipeReducer.errorMessage
});

const mapDispatchToProps = dispatch => (
  bindActionCreators({
    getUserRecipe,
    addRecipeAction,
    deleteRecipeAction,
    editRecipeAction,
    saveImageToCloudAction,
    setEditRecipeIdAction
  }, dispatch)
);

export default connect(mapStateToProps, mapDispatchToProps)(UserRecipes);
