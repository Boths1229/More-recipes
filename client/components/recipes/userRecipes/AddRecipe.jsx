import React from 'react';
import AddModal from '../../Include/AddModal';

const AddRecipe = (props) => {
  const { onChange, addRecipe, value } = props;
  return (
    <div className="row header-nav ">
      <div className="col-sm-6 col-6">
        <p className="pt-3 font">Add New recipe</p>
      </div>
      <div className="col-sm-6 col-6 pt-2 pb-2">
        <button
          type="button"
          className="btn rfloat mt-2"
          data-toggle="modal"
          data-target="#exampleModal"
        >
          <i className="fa fa-plus-square-o" aria-hidden="true" />
        </button>
        <AddModal
          onChange={props.onChange}
          addRecipe={props.addRecipe}
          value={props.value}
        />
      </div>
    </div>
  );
};

export default AddRecipe;
