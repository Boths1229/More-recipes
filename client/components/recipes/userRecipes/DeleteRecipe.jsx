import React from 'react';
import PropTypes from 'prop-types';

/**
 * @description functional component to render delete button for recipe
 *
 * @param {props} props
 *
 * @returns {JSX} return JSX
 */
const DeleteRecipe = (props) => {
  const { recipeId } = props;
  return (
    <div>
      <button
        type="button"
        className="btn btn-danger ml-3"
        data-toggle="modal"
        data-target={`#deleteModal-${recipeId}`}
        data-whatever="@getbootstrap"
      >
        <i className="fa fa-times" aria-hidden="true" />
      </button>
    </div>
  );
};

DeleteRecipe.propTypes = {
  recipeId: PropTypes.number.isRequired
};

export default DeleteRecipe;
