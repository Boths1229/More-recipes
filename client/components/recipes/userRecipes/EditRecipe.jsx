import React from 'react';
import PropTypes from 'prop-types';

/**
 * @description function component to render edit recipe button
 *
 * @param {props} props
 *
 * @returns {JSX} return JSX
 */
const EditRecipe = (props) => {
  const {
    recipeId, showRecipeDetails, cardId,
  } = props;
  return (
    <div>
      <button
        type="button"
        className="btn"
        data-toggle="modal"
        data-target={`#${cardId}`}
        data-whatever="@getbootstrap"
        onClick={() => showRecipeDetails(recipeId)}
      >
        <i className="fa fa-pencil-square-o" aria-hidden="true" />
      </button>
    </div>
  );
};

EditRecipe.propTypes = {
  recipeId: PropTypes.number.isRequired,
  showRecipeDetails: PropTypes.func.isRequired,
  cardId: PropTypes.number.isRequired
};

export default EditRecipe;
