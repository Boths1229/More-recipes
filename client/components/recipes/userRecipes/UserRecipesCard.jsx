import React from 'react';
import PropTypes, { any } from 'prop-types';
import EditRecipe from './EditRecipe';
import DeleteRecipe from './DeleteRecipe';
import EditModal from '../../Include/EditModal';
import DeleteModal from '../../Include/DeleteModal';

const UserRecipesCard = (props) => {
  const {
    recipeData,
    showRecipeDetails,
    cardId,
    deleteRecipe,
    editRecipe,
    value,
    onChange,
    saveImageToCloud
  } = props;
  const {
    id, name, description, image
  } = recipeData;
  return (
    <div className="col-sm-6 col-md-6 col-lg-4 pt-4">
      <div className="card">
        <img className="card-img-top img-max" src={image} alt="" />
        <div className="card-body">
          <a href="">
            <h4 className="card-title">{name}</h4>
          </a>
          <p className="card-text pt-2">
            {description}
          </p>
        </div>
        <div className="card-footer card-tile">
          <div className="row">
            <div className="col-xs-6 ml-3">
              <EditRecipe
                recipeId={id}
                cardId={cardId}
                showRecipeDetails={showRecipeDetails}
              />
              <EditModal
                cardId={cardId}
                editRecipe={editRecipe}
                value={value}
                onChange={onChange}
                recipeId={id}
                recipeData={recipeData}
                saveImageToCloud={saveImageToCloud}
              />
            </div>
            <div className="col-xs-6">
              <DeleteRecipe />
              <DeleteModal
                deleteRecipe={deleteRecipe}
                recipeId={id}
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

UserRecipesCard.propTypes = {
  recipeData: PropTypes.objectOf(any).isRequired,
  deleteRecipe: PropTypes.func.isRequired,
  editRecipe: PropTypes.func.isRequired,
  // value: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  saveImageToCloud: PropTypes.func.isRequired
};

export default UserRecipesCard;
