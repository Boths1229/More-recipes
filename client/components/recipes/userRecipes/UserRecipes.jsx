import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes, { any } from 'prop-types';
import { filter } from 'lodash';
import getUserRecipe from '../../../actionController/getUserRecipe';
import addRecipeAction from '../../../actionController/addRecipe';
import saveImageToCloudAction from '../../../actionController/saveImageToCloud';
import AddRecipe from './AddRecipe';
import deleteRecipeAction from '../../../actionController/deleteRecipe';
import editRecipeAction from '../../../actionController/editRecipe';
import UserRecipesCard from './UserRecipesCard';
import Header from '../../common/Header';
import Footer from '../../common/Footer';


/**
 * @class UserRecipes
 * @description UserRecipes
 */
class UserRecipes extends Component {
  /**
   * @description create an instance of UserRecipes
   * @param {*} props
   */
  constructor(props) {
    super(props);
    this.state = {
      recipeData: [],
      isFetched: false,
      recipe: [],
      name: '',
      isChanged: false
    };
    this.handleChange = this.handleChange.bind(this);
    this.addRecipe = this.addRecipe.bind(this);
    this.deleteRecipe = this.deleteRecipe.bind(this);
    this.saveImageToCloud = this.saveImageToCloud.bind(this);
    this.editRecipe = this.editRecipe.bind(this);
    this.handleShowRecipe = this.handleShowRecipe.bind(this);
  }
  /**
   * @description check the state of isFetched and call the get recipe action
   * @param {*} props
   * @return {*} return all recipes
   */
  componentDidMount() {
    if (this.props.recipes.isFetched === false) {
      this.props.getUserRecipe(this.props.userId);
    }
  }

  /**
   * @description update the state of user recipes
   * @param {*} nextProps
   * @return {*} updated user recipe
   */
  componentWillReceiveProps(nextProps) {
    const { recipeImageUrl, recipes } = nextProps;
    const { recipeData, isFetched } = recipes;
    if (recipeData !== this.props.recipes.recipeData) {
      this.setState({
        recipeData,
        isFetched
      });
    }
    if (recipeImageUrl !== this.props.recipeImageUrl) {
      this.setState({
        image: recipeImageUrl
      });
    }
  }
  /**
   * @description ui
   * @param {recipeData id} recipeData id
   * @return {null} l
   */
  editRecipe(id, recipeData) {
    console.log("@@@@@@@@@@@@@@", recipeData);
    event.preventDefault();
    this.props.editRecipeAction(id, recipeData);
  }

  /**
   * @description ui
   * @param {id} id
   * @return {null} l
   */
  deleteRecipe(id) {
    this.props.deleteRecipeAction(id);
  }

  /**
   * @description ui
   * @param {event} event
   * @return {null} l
   */
  handleChange(event) {
    const { name, value } = event.target;
    this.setState({
      [name]: value,
      isChanged: true
    });
  }

  handleShowRecipe(recipeId) {
    const recipe = filter(this.state.recipeData, recipe => recipe.id === recipeId)
    this.setState({
      recipe
    })
  }

  /**
   * @description ui
   * @param {event} event
   * @return {null} l
   */
  addRecipe(event) {
    event.preventDefault();
    const {
      name,
      image,
      description,
      ingredient
    } = this.state;
    const newRecipe = {
      name,
      image,
      description,
      ingredient
    };
    this.props.addRecipeAction(newRecipe);
    this.setState({
      name: '', image: '', description: '', ingredient: ''
    });
  }

  /**
   * @description ui
   * @param {event} event
   * @return {null} l
   */
  saveImageToCloud(event) {
    const image = event.target.files[0];
    if (image) {
      this.props.saveImageToCloudAction(image);
    }
  }
  /**
   * @description render user recipes
   * @return {*} wfdgsfd
   */
  render() {
    let renderUserRecipes = <h1>No recipes in your catalog</h1>;
    if (this.state.isFetched) {
      renderUserRecipes = this.state.recipeData.map((recipeData, key) => (
        <UserRecipesCard
          key={key}
          cardId={key}
          recipeData={recipeData}
          showRecipeDetails={this.handleShowRecipe}
          deleteRecipe={this.deleteRecipe}
          editRecipe={this.editRecipe}
          addRecipe={this.addRecipe}
          value={this.state}
          onChange={this.handleChange}
          saveImageToCloud={this.saveImageToCloud}
        />
      ));
    }
    return (
      <div>
        <Header />
        <div className="container">
          <div className="top-nav-bar">
            <AddRecipe
              value={this.state}
              onChange={this.handleChange}
              addRecipe={this.addRecipe}
              saveImageToCloud={this.saveImageToCloud}
            />

          </div>
          <div className="row">
            {this.state.isFetched && renderUserRecipes}
          </div>
          <Footer />
        </div>
      </div>
    );
  }
}
UserRecipes.propTypes = {
  getUserRecipe: PropTypes.func.isRequired,
  addRecipeAction: PropTypes.func.isRequired,
  deleteRecipeAction: PropTypes.func.isRequired,
  editRecipeAction: PropTypes.func.isRequired,
  recipes: PropTypes.objectOf(any).isRequired,
  // recipeImageUrl: PropTypes.string.isRequired,
  saveImageToCloudAction: PropTypes.func.isRequired,
  // userId: PropTypes.objectOf(any).isRequired,
};

const mapStateToProps = state => ({
  recipes: state.getUserRecipeReducer[0],
  userId: state.signinReducer[0].userData.id,
  recipeImageUrl: state.saveImageToCloud[0].image
});

export default connect(mapStateToProps, {
  getUserRecipe,
  addRecipeAction,
  deleteRecipeAction,
  editRecipeAction,
  saveImageToCloudAction
})(UserRecipes);
