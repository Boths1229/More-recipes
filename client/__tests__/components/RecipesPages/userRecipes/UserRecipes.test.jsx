import React from 'react';
import { shallow } from 'enzyme';
import { UserRecipes } from '../../../../components/recipePages/userRecipes/UserRecipes';

describe('Recipe details page component', () => {
  it('matches snapshot', () => {
    const props = {
      history: {
        push: jest.fn()
      }
    };
    const output = shallow(<UserRecipes {...props} />);
    expect(output).toMatchSnapshot();
  });
  describe('instance and lifecycle methods', () => {
    const props = {
      getUserRecipe: jest.fn(),
      addRecipeAction: jest.fn(),
      editRecipeAction: jest.fn(),
      clearImageAction: jest.fn(),
      deleteRecipeAction: jest.fn(),
      setEditRecipeIdAction: jest.fn(),
      saveImageToCloudAction: jest.fn(),
      history: {
        push: jest.fn()
      },
      addRecipeResponse: {},
      recipes: {
        recipeData: ''
      }
    };
    it('calls componentWillRecieveProps', () => {
      const wrapper = shallow(<UserRecipes {...props} />);
      const cwrp = jest.spyOn(wrapper.instance(), 'componentWillReceiveProps');
      wrapper.setProps({ addRecipeResponse: { errorMessage: 'error message' }, });
      expect(cwrp).toHaveBeenCalled();
    });
    it('handleChange', () => {
      const wrapper = shallow(<UserRecipes {...props} />);
      const setState = jest.spyOn(wrapper.instance(), 'setState');
      wrapper.instance().handleChange({ target: { name: 'description', value: '' } });
      expect(setState).toHaveBeenCalledWith({ description: '', isChanged: true });
    });
    it('handleAddRecipe', () => {
      const wrapper = shallow(<UserRecipes {...props} />);
      wrapper.instance().addRecipe({ preventDefault: jest.fn() });
      expect(props.addRecipeAction).toBeCalled();
    });
    it('handles clearImageState', () => {
      const wrapper = shallow(<UserRecipes {...props} />);
      wrapper.instance().clearImageState();
      expect(props.clearImageAction).toBeCalled();
    });
    it('handles deleteRecipe', () => {
      const wrapper = shallow(<UserRecipes {...props} />);
      wrapper.instance().deleteRecipe();
      expect(props.deleteRecipeAction).toBeCalled();
    });
    it('handles setEditRecipeId', () => {
      const wrapper = shallow(<UserRecipes {...props} />);
      wrapper.instance().setEditRecipeId();
      expect(props.setEditRecipeIdAction).toBeCalled();
    });
    it('handles handlePaginationChange', () => {
      const wrapper = shallow(<UserRecipes {...props} />);
      wrapper.instance().handlePaginationChange({ recipes: {} });
      expect(props.getUserRecipe).toBeCalled();
    });
    it('handles editRecipe', () => {
      const wrapper = shallow(<UserRecipes {...props} />);
      wrapper.instance().editRecipe();
      expect(props.editRecipeAction).toBeCalled();
    });
    it('handles saveImageToCloud', () => {
      const wrapper = shallow(<UserRecipes {...props} />);
      wrapper.instance().saveImageToCloud({ target: { files: [{ image: 'recipe' }] } });
      expect(props.saveImageToCloudAction).toBeCalled();
    });
  });
});
