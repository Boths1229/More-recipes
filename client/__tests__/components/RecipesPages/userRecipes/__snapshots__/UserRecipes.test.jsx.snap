// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe details page component matches snapshot 1`] = `
<div>
  <Connect(Header) />
  <div
    className="container"
  >
    <div
      className="top-nav-bar"
    >
      <AddRecipeButton
        addRecipe={[Function]}
        onChange={[Function]}
        saveImageToCloud={[Function]}
        value={
          Object {
            "description": "",
            "ingredient": "",
            "isChanged": false,
            "isFetched": false,
            "name": "",
            "recipeData": Array [],
            "responseMessage": "",
          }
        }
      />
    </div>
    <div>
      <div
        className="row"
      >
        <RecipeGrid
          addRecipe={[Function]}
          clearImageState={[Function]}
          deleteRecipe={[Function]}
          description=""
          editRecipe={[Function]}
          ingredient=""
          isChanged={false}
          isFetched={false}
          name=""
          onChange={[Function]}
          recipeData={Array []}
          responseMessage=""
          setEditRecipeId={[Function]}
          showActionButton={true}
          value={
            Object {
              "description": "",
              "ingredient": "",
              "isChanged": false,
              "isFetched": false,
              "name": "",
              "recipeData": Array [],
              "responseMessage": "",
            }
          }
        />
      </div>
    </div>
    <PaginationBoxView
      activeClassName="active"
      breakClassName="page-link"
      breakLabel={
        <a
          href=""
        >
          ...
        </a>
      }
      containerClassName="pagination justify-content-center"
      disableInitialCallback={false}
      disabledClassName="disabled"
      marginPagesDisplayed={3}
      nextClassName="page-item"
      nextLabel="Next"
      nextLinkClassName="page-link"
      onPageChange={[Function]}
      pageClassName="page-item"
      pageCount={10}
      pageLinkClassName="page-link"
      pageRangeDisplayed={2}
      previousClassName="page-item"
      previousLabel="Previous"
      previousLinkClassName="page-link"
      subContainerClassName="pages pagination"
    />
    <Footer />
  </div>
</div>
`;
